src/scripting_api/util.js

```diff
@@ -20,11 +20,724 @@ class Util extends PDFObject {
+      if (p2) {
+        for (const c of p2) {
+          switch (c) {
+            case "+":
+              cFlags = cFlags | PLUS;
+              break;
+            case " ":
+              cFlags = cFlags | SPACE;
+              break;
+            case "0":
+              cFlags = cFlags | ZERO;
+              break;
+            case "#":
+              cFlags = cFlags | HASH;
+              break;
+          }
+        }
+      }
+
+      let nWidth;
+      if (p3) {
+        nWidth = parseInt(p3);
+      }
+
+      let intPart = Math.trunc(arg);
+
+      if (cConvChar === "x") {
+        let hex = Math.abs(intPart).toString(16).toUpperCase();
+        if (nWidth !== undefined) {
+          hex = hex.padStart(nWidth, cFlags & ZERO ? "0" : " ");
+        }
+        if (cFlags & HASH) {
+          hex = `0x${hex}`;
+        }
+        return hex;
+      }
+
+      const nDecSep = p1 ? p1.substring(1) : "0";
+      let nPrecision;
+      if (p4) {
+        nPrecision = parseInt(p4.substring(1));
+      }
+
+      const separators = {
+        0: [",", "."],
+        1: ["", "."],
+        2: [".", ","],
+        3: ["", ","],
+      };
+      const [thousandSep, decimalSep] =
+        nDecSep in separators ? separators[nDecSep] : ["'", "."];
+
+      let decPart = "";
+      if (cConvChar === "f") {
+        if (nPrecision !== undefined) {
+          decPart = (arg - intPart).toFixed(nPrecision);
+        } else {
+          decPart = (arg - intPart).toString();
+        }
+        if (decPart.length > 2) {
+          decPart = `${decimalSep}${decPart.substring(2)}`;
+        } else if (cFlags & HASH) {
+          decPart = ".";
+        } else {
+          decPart = "";
+        }
+      }
+
+      let prefix = "";
+      if (intPart < 0) {
+        prefix = "-";
+        intPart = -intPart;
+      } else if (cFlags & PLUS) {
+        prefix = "+";
+      } else if (cFlags & SPACE) {
+        prefix = " ";
+      }
+
+      if (thousandSep && intPart >= 1000) {
+        const buf = [];
+        while (true) {
+          buf.push((intPart % 1000).toString().padStart(3, "0"));
+          intPart = Math.trunc(intPart / 1000);
+          if (intPart < 1000) {
+            buf.push(intPart.toString());
+            break;
+          }
+        }
+        intPart = buf.reverse().join(thousandSep);
+      } else {
+        intPart = intPart.toString();
+      }
+
+      let n = `${intPart}${decPart}`;
+      if (nWidth !== undefined) {
+        n = n.padStart(nWidth - prefix.length, cFlags & ZERO ? "0" : " ");
+      }
+
+      return `${prefix}${n}`;
+    });
+  }
+
+  iconStreamFromIcon() {
+    /* Not implemented */
+  }
+
+  printd(cFormat, cDate, bXFAPicture = false) {
+    if (bXFAPicture) {
+      return this._printdXFAPicture(cFormat, cDate);
+    }
+
+    switch (cFormat) {
+      case "0":
+        return this.printd("D:yyyymmddHHMMss", cDate);
+      case "1":
+        return this.printd("yyyy.mm.dd HH:MM:ss", cDate);
```

Comment:
Unless I'm missing something here, these formats don't match the examples at https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/js_api_reference.pdf#page=718&zoom=auto,-207,541. Value 1 starts with `D:` for example, while this format doesn't, and the one above can also have a timezone offset specifier. The one below is yyyy/mm/dd in the spec.

Reply:
The spec is pretty quite strange here:
 - 0 & 1 are the same
 - in the `cFormat` there is nothing related to the timezone offset
 So I took the format in pdfium:
https://pdfium.googlesource.com/pdfium/+/refs/heads/master/fxjs/cjs_util.cpp#187

But in looking at 7.9.4 in pdf spec a date has the format `D:YYYYMMDDHHmmSSOHH'mm` so it could have an offset specifier.
And maybe "Universal" (1) should be `YYYY-MM-DDTHH:mm:SSZHH(Z|[+-]]HHmm)`.
Wdyt ?

Reply:
I just checked in Acrobat (Mac version) and the display is exactly the same as the one defined in pdfium source.
So I guess the js specs are incorrect here.

Reply:
Ugh, Adobe's specs are indeed sometimes confusing, incomplete or simply don't match their reference implementation (Adobe Reader/Acrobat), so unfortunately I'm not really surprised. Thank you for figuring this all out!